### PR TITLE
Feat: Add ability to provide backup GPUs which containers can be scheduled on 

### PR DESCRIPTION
**Pull request by jsun-m**
This update introduces a new feature that allows users to specify multiple GPUs for fallback purposes. In scenarios where the primary GPU becomes unavailable due to low capacity or failure, the system will now automatically fallback to the specified alternative GPUs.

#### Key Changes:
•	GPU Parameter Flexibility:
•	The gpu parameter in all abstractions now accepts both a single GPU as a string and a list of GPUs as strings.
•	This change enhances flexibility in GPU management, enabling users to specify a prioritized list of fallback GPUs.

#### Usage Examples:
1.	Single GPU Specification:
```
@function(
    gpu="A10G",
)
def test():
    ...
```
2.	Multiple GPU Specification:
```
@function(
    gpu=["A10G", "T4"]
)
def test():
    ...
```

#### Priority Handling:
The order of GPUs specified in the list determines their priority. The GPU at the earliest index in the list is given the highest priority, allowing for efficient fallback management.

This enhancement aims to improve resource allocation and ensure more reliable GPU utilization within our system.

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -237,6 +259,7 @@ func (s *Scheduler) scheduleRequest(worker *types.Worker, request *types.Contain
 	go s.schedulerMetrics.CounterIncContainerScheduled(request)
 	go s.eventRepo.PushContainerScheduledEvent(request.ContainerId, worker.Id, request)
 
+	request.ActualGpu = worker.Gpu
```
Comment (luke-lombardi) [inline pkg/scheduler/scheduler.go:262]: this is super confusing

File: pkg/types/scheduler.go [hunk 2]
```diff
@@ -67,24 +67,28 @@ type ContainerState struct {
...
+	ContainerId string   `json:"container_id"`
+	EntryPoint  []string `json:"entry_point"`
+	Env         []string `json:"env"`
+	Cpu         int64    `json:"cpu"`
+	Memory      int64    `json:"memory"`
+	Gpu         string   `json:"gpu"`
+	BackupGpus  []string `json:"backup_gpus"`
+	// TODO: Ideally Gpu and BackupGpus should be merged into a single field
+	// However we need to do it in a backward compatible way
+	ActualGpu        string          `json:"actual_gpu"` // The actual GPU that was chosen for the container
```
Comment (luke-lombardi) [inline pkg/types/scheduler.go:79]: this field shouldn't be necessary

File: pkg/abstractions/container/container.go [hunk 3]
```diff
@@ -158,12 +159,21 @@ func (cs *CmdContainerService) ExecuteCommand(in *pb.CommandExecutionRequest, st
 		gpuCount = 1
 	}
 
+	gpuTypes := strings.Split(string(stubConfig.Runtime.Gpu), ",")
```
Comment (luke-lombardi) [inline pkg/abstractions/container/container.go:162]: can you consolidate this logic into a helper function or something?

**Review by luke-lombardi — CHANGES_REQUESTED** [review]

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -161,23 +161,37 @@ func (s *Scheduler) Stop(containerId string) error {
...
 
-func (s *Scheduler) getController(request *types.ContainerRequest) (WorkerPoolController, error) {
-	var ok bool
-	var workerPool *WorkerPool
+func (s *Scheduler) getControllers(request *types.ContainerRequest) ([]WorkerPoolController, error) {
+	controllers := []WorkerPoolController{}
 
 	if request.PoolSelector != "" {
-		workerPool, ok = s.workerPoolManager.GetPool(request.PoolSelector)
+		wp, ok := s.workerPoolManager.GetPool(request.PoolSelector)
```
Comment (jsun-m) [inline pkg/scheduler/scheduler.go:168]: will look into simplifying this.

**Comment by luke-lombardi** [comment]
@jsun-m I'm still reviewing this, but can you rename this PR and add more context to the PR description?

File: pkg/abstractions/container/container.go [hunk 5]
```diff
@@ -154,16 +154,21 @@ func (cs *CmdContainerService) ExecuteCommand(in *pb.CommandExecutionRequest, st
 	env = append(secrets, env...)
 
 	gpuCount := 0
-	if stubConfig.Runtime.Gpu != "" {
+	if len(stubConfig.Runtime.Gpus) > 0 {
 		gpuCount = 1
 	}
 
+	gpuRequest := types.GpuTypesToStrings(stubConfig.Runtime.Gpus)
+	if stubConfig.Runtime.Gpu != "" {
```
Comment (luke-lombardi) [inline pkg/abstractions/container/container.go:162]: this is the backwards compatibility check correct?

File: pkg/repository/container_redis.go [hunk 6]
```diff
@@ -161,6 +161,39 @@ func (cr *ContainerRedisRepository) UpdateContainerStatus(containerId string, st
 	return nil
 }
 
+func (cr *ContainerRedisRepository) UpdateContainerGPU(containerId string, gpuType string) error {
```
Comment (luke-lombardi) [inline pkg/repository/container_redis.go:164]: maybe UpdateAssignedContainerGPU?

File: pkg/types/scheduler.go [hunk 7]
```diff
@@ -74,18 +74,19 @@ type ContainerRequest struct {
...
 	ImageId          string          `json:"image_id"`
 	StubId           string          `json:"stub_id"`
 	WorkspaceId      string          `json:"workspace_id"`
 	Workspace        Workspace       `json:"workspace"`
-	Stub             StubWithRelated `json:"stub"`
 	Timestamp        time.Time       `json:"timestamp"`
 	Mounts           []Mount         `json:"mounts"`
 	RetryCount       int             `json:"retry_count"`
 	PoolSelector     string          `json:"pool_selector"`
+	Stub             StubWithRelated `json:"stub"`
```
Comment (luke-lombardi) [inline pkg/types/scheduler.go:89]: why did you move this? I thought it sort of made sense next to Workspace

File: sdk/src/beta9/abstractions/base/runner.py [hunk 8]
```diff
@@ -272,6 +272,14 @@ def _sync_content(
         except Exception as e:
             terminal.warn(str(e))
 
+    def _parse_gpu(self, gpu: Union[GpuTypeAlias, List[GpuTypeAlias]]) -> str:
```
Comment (luke-lombardi) [inline sdk/src/beta9/abstractions/base/runner.py:275]: does this handle type assertions / validation on the input?

File: pkg/scheduler/scheduler.go [hunk 9]
```diff
@@ -252,12 +278,35 @@ func filterWorkersByPoolSelector(workers []*types.Worker, request *types.Contain
...
+		}
+
+		if len(gpuRequestsMap) > 0 {
+			// Validate GPU resource availability
+			priorityModifier, validGpu := gpuRequestsMap[worker.Gpu]
+			if !validGpu || worker.FreeGpuCount < request.GpuCount {
+				continue
+			}
+
+			worker.Priority -= int32(priorityModifier) // This will account for the preset priorities for the pool type as well as the order of the GPU requests
```
Comment (luke-lombardi) [inline pkg/scheduler/scheduler.go:306]: hm interesting. will this always work? i think it'll only work if priorities are just increasing by 1?

**Review by luke-lombardi — CHANGES_REQUESTED** [review]

File: pkg/abstractions/container/container.go (hunk 5 above)
Comment (jsun-m) [reply pkg/abstractions/container/container.go:162]: yes

File: pkg/types/scheduler.go (hunk 7 above)
Comment (jsun-m) [reply pkg/types/scheduler.go:89]: not sure why i moved it. moving back

File: pkg/scheduler/scheduler.go (hunk 9 above)
Comment (jsun-m) [reply pkg/scheduler/scheduler.go:306]: right for now it will work well if priorities are incremental. Once we think of more complex priorities, we would need to update this scoring function. But this additive/subtractive behavior works well with the default priorities.

File: sdk/src/beta9/abstractions/base/runner.py (hunk 8 above)
Comment (jsun-m) [reply sdk/src/beta9/abstractions/base/runner.py:290]: yes

File: pkg/repository/container_redis.go (hunk 6 above)
Comment (jsun-m) [reply pkg/repository/container_redis.go]: done

**Review by luke-lombardi — CHANGES_REQUESTED** [review]
Can you add a note about that here?

File: pkg/scheduler/scheduler.go [hunk 10]
```diff
@@ -161,23 +161,36 @@ func (s *Scheduler) Stop(containerId string) error {
 	return nil
 }
 
-func (s *Scheduler) getController(request *types.ContainerRequest) (WorkerPoolController, error) {
-	var ok bool
-	var workerPool *WorkerPool
+func (s *Scheduler) getControllers(request *types.ContainerRequest) ([]WorkerPoolController, error) {
+	controllers := []WorkerPoolController{}
+	combinedRequestedGpu := request.GpuRequest
```
Comment (luke-lombardi) [inline pkg/scheduler/scheduler.go:166]: Can you come up with a better name for this car

**Review by luke-lombardi — CHANGES_REQUESTED** [review]
two more comments

**Review by luke-lombardi — APPROVED** [review]
